File: neetcode150/17_math_and_geometry/2013_detect_squares.cpp
#include <bits/stdc++.h>
#include <chrono>
using namespace std;
int main(){
    auto t=chrono::high_resolution_clock::now();
    unordered_map<int, unordered_map<int, int>> ptsCount; vector<pair<int, int>> pts;
    auto add = [&](int x, int y){ ptsCount[x][y]++; pts.push_back({x, y}); };
    auto count = [&](int px, int py){ int res = 0; for(auto& p : pts){ int x = p.first, y = p.second; if(abs(py - y) != abs(px - x) || x == px || y == py) continue; res += ptsCount[x][py] * ptsCount[px][y]; } return res; };
    add(3,10); add(11,2); add(3,2); int res = count(11,10);
    cout<<"var1: "<<res<<endl;
    auto e=chrono::high_resolution_clock::now();
    cout<<"var1 time="<<chrono::duration_cast<chrono::nanoseconds>(e-t).count()/1000.0<<"us"<<endl;
}
// var1 mem = {} and time = {}

File: neetcode150/17_math_and_geometry/2013_detect_squares.py
"""LeetCode 2013: Detect Squares - Medium"""
import time, tracemalloc
from collections import defaultdict
from typing import List
# Variation 1: Hash map of points
class DetectSquares_v1:
    def __init__(self):
        self.ptsCount = defaultdict(int)
        self.pts = []
    def add(self, point: List[int]) -> None:
        self.ptsCount[tuple(point)] += 1
        self.pts.append(point)
    def count(self, point: List[int]) -> int:
        res = 0; px, py = point
        for x, y in self.pts:
            if (abs(py - y) != abs(px - x)) or x == px or y == py: continue
            res += self.ptsCount[(x, py)] * self.ptsCount[(px, y)]
        return res
if __name__ == "__main__":
    tracemalloc.start(); t0=time.perf_counter()
    obj = DetectSquares_v1()
    obj.add([3,10]); obj.add([11,2]); obj.add([3,2])
    r = obj.count([11,10])
    t1=time.perf_counter(); mem=tracemalloc.get_traced_memory()[1]; tracemalloc.stop()
    print(f"var1: res={r}, mem={mem}bytes, time={(t1-t0)*1e6:.2f}us")
# var1 mem = {} and time = {}
